synapse/module_api/__init__.py

```diff
@@ -293,6 +296,62 @@ async def complete_sso_login_async(
             registered_user_id, request, client_redirect_url,
         )
 
+    async def create_and_send_event_into_room(
+        self,
+        sender: str,
+        room_id: str,
+        event_type: str,
+        content: JsonDict,
+        state_key: Optional[str] = None,
+    ) -> EventBase:
+        """Create and send an event into a room.
+
+        Args:
+            sender: The user ID of the event sender.
+            room_id: The ID of the room to send the event into.
+            event_type: The type of the event.
+            content: The event content.
+            state_key: The event's state key. None if this is not a state event.
+
+        Returns:
+            The event that was sent.
+
+        Raises:
+            SynapseError if the event was not allowed.
+        """
+        if event_type == EventTypes.Member:
+            raise Exception(
+                "ThirdPartyEventRules modules are not currently able to send "
+                "m.room.member events"
+            )
+
+        # Build event dictionary
+        event_dict = {
+            "type": event_type,
+            "content": content,
+            "room_id": room_id,
+            "sender": sender,
+        }
+
+        # Check whether this is intended to be a state event
+        if state_key is not None:
+            event_dict["state_key"] = state_key
+
+        # Create a requester object
+        requester = create_requester(sender)
+
+        # Create and send the event
+        event, _ = await self._event_creation_handler.create_and_send_nonmember_event(
+            requester,
+            event_dict,
+            ratelimit=False,
+            ignore_shadow_ban=True,
+            ignore_spam_check=True,
+            ignore_third_party_event_rules=True,
```

Comment:
is this to avoid loops? I think it would be better for the third-party-rules module to be responsible for avoiding such loops.

Reply:
apart from being annoying to thread this condition through, we have no guarantee that the only module using this api endpoint is the `ThirdPartyEventRules` module. Other modules can exist.

Reply:
It was mostly to avoid the risk of other modules, or other parts of the codebase, trying to send events that don't originate from user action getting caught out by `ThirdPartyEventRules`.

But I'm torn, as I don't want `ThirdPartyEventRules` from being unable to detect internally-sent events from coming through, as that limits its usefulness.

Reply:
The same could be said about the `ignore_spam_check` option really.

Reply:
We've concluded in #synapse-dev to strip both of these, and let modules check events for others. If a conflict arises, that falls on the responsibility of the one who installed conflicting modules - aka the server admin and module devs.